Mathlib/LinearAlgebra/TensorProduct/Quotient.lean

```diff
@@ -0,0 +1,241 @@
+  Submodule.Quotient.equiv _ _ (LinearEquiv.refl _ _) (by
+    simp only [Submodule.map_sup]
+    erw [Submodule.map_id, Submodule.map_id]
+    simp only [sup_eq_left]
+    rw [map_range_eq_span_tmul, map_range_eq_span_tmul]
+    aesop)
+
+@[simp]
+lemma quotientTensorEquiv_apply_tmul_mk (m : Submodule R M) (x : M) (y : N) :
+    quotientTensorEquiv N m (Submodule.Quotient.mk x ⊗ₜ[R] y) =
+    Submodule.Quotient.mk (x ⊗ₜ y) :=
+  rfl
+
+@[simp]
+lemma quotientTensorEquiv_symm_apply_mk_tmul (m : Submodule R M) (x : M) (y : N) :
+    (quotientTensorEquiv N m).symm (Submodule.Quotient.mk (x ⊗ₜ y)) =
+    Submodule.Quotient.mk x ⊗ₜ[R] y :=
+  rfl
+
+variable (M) in
+/--
+Let `M, N` be `R`-modules, `n ≤ N` be an `R`-submodule. Then we have a linear isomorphism between
+tensor products of the quotient and the quotient of the tensor product:
+`M ⊗[R] (N ⧸ n) ≃ₗ[R] (M ⊗[R] N) ⧸ (M ⊗ n)`.
+-/
+noncomputable def tensorQuotientEquiv (n : Submodule R N) :
+    M ⊗[R] (N ⧸ (n : Submodule R N)) ≃ₗ[R]
+    (M ⊗[R] N) ⧸ (LinearMap.range (map (LinearMap.id : M →ₗ[R] M) n.subtype)) :=
+  congr ((Submodule.quotEquivOfEqBot _ rfl).symm) (LinearEquiv.refl _ _)  ≪≫ₗ
+  quotientTensorQuotientEquiv (⊥ : Submodule R M) n ≪≫ₗ
+  Submodule.Quotient.equiv _ _ (LinearEquiv.refl _ _) (by
+    simp only [Submodule.map_sup]
+    erw [Submodule.map_id, Submodule.map_id]
+    simp only [sup_eq_right]
+    rw [map_range_eq_span_tmul, map_range_eq_span_tmul]
+    aesop)
+
+@[simp]
+lemma tensorQuotientEquiv_apply_mk_tmul (n : Submodule R N) (x : M) (y : N) :
+    tensorQuotientEquiv M n (x ⊗ₜ[R] Submodule.Quotient.mk y) =
+    Submodule.Quotient.mk (x ⊗ₜ y) :=
+  rfl
+
+@[simp]
+lemma tensorQuotientEquiv_symm_apply_tmul_mk (n : Submodule R N) (x : M) (y : N) :
+    (tensorQuotientEquiv M n).symm (Submodule.Quotient.mk (x ⊗ₜ y)) =
+    x ⊗ₜ[R] Submodule.Quotient.mk y :=
+  rfl
+
+variable (M) in
+/-- Left tensoring a module with a quotient of the ring is the same as
+quotienting that module by the corresponding submodule. -/
+noncomputable def quotTensorEquivQuotSMul (I : Ideal R) :
+    ((R ⧸ I) ⊗[R] M) ≃ₗ[R] M ⧸ (I • (⊤ : Submodule R M)) :=
+  quotientTensorEquiv M I ≪≫ₗ
+  Submodule.Quotient.equiv (M := R ⊗[R] M) (N := M) (f := TensorProduct.lid R M) (hf := rfl) ≪≫ₗ
+  Submodule.Quotient.equiv _ _ (LinearEquiv.refl R M) (by
+    erw [Submodule.map_id]
+    rw [TensorProduct.map_range_eq_span_tmul, Submodule.map_span]
+    refine le_antisymm (Submodule.span_le.2 ?_) (Submodule.map₂_le.2 ?_)
+    · rintro _ ⟨_, ⟨r, m, rfl⟩, rfl⟩
+      simp only [Submodule.coe_subtype, LinearMap.id_coe, id_eq, lid_tmul, SetLike.mem_coe]
+      apply Submodule.apply_mem_map₂ <;> aesop
+    · rintro r hr m -
+      simp only [Submodule.coe_subtype, LinearMap.id_coe, id_eq, Subtype.exists, exists_prop,
+        LinearMap.lsmul_apply]
+      refine Submodule.subset_span ?_
+      simp only [Set.mem_image, Set.mem_setOf_eq]
+      exact ⟨r ⊗ₜ m, ⟨r, hr, m, rfl⟩, rfl⟩)
```

Comment:
This declaration is moved from RightExactness, but the construction becomes much longer. Is it for the purpose of avoiding `rTensor.equiv`? Why is this a price worth paying?
```
  (rTensor.equiv M (exact_subtype_mkQ I) I.mkQ_surjective).symm.trans <|
    Submodule.Quotient.equiv _ _ (TensorProduct.lid R M) <|
      Eq.trans (LinearMap.range_comp _ _).symm <|
        Eq.trans ((Submodule.topEquiv.lTensor I).range_comp _).symm <|
          Eq.symm <| Eq.trans (map₂_eq_range_lift_comp_mapIncl _ _ _) <|
            congrArg _ (TensorProduct.ext' (fun _ _ => rfl))
```

Reply:
I consider `(R ⧸ I) ⊗ M ≃ IM` as a special case of [TensorProduct.quotientTensorQuotientEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/Quotient.html#TensorProduct.quotientTensorQuotientEquiv)

Reply:
I think the definition now and the definition previously are still defeq, for all the other lemmas about `quotTensorEquivQuotSMul` works verbatim

Reply:
Since nothing breaks, I'm importing RightExactness in my PR and reverting to the old construction using `rTensor.equiv`. It might be possible to construct [TensorProduct.quotientTensorQuotientEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/Quotient.html#TensorProduct.quotientTensorQuotientEquiv) using quotientTensorEquiv and tensorQuotientEquiv together with [Submodule.quotientQuotientEquivQuotientSup](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Isomorphisms.html#Submodule.quotientQuotientEquivQuotientSup), but I probably won't attempt additional refactor within this file in the near future.